psrqpy/search.py

```diff
@@ -15,686 +15,2378 @@
+        """
+
+        self.parse_assoc()      # parse the association parameter
+        self.parse_type()       # parse the type parameter
+        self.parse_bincomp()    # parse the binary companion parameter
+
+    def parse_assoc(self):
+        """
+        Parse default string representing source associations, extracting (first) value
+        and reference. Multiple values and references currently not supported.
+        """
+
+        if 'ASSOC' not in self.columns:
+            return
+
+        # Save original parameter in new column
+        ASSOCorig = self.catalogue['ASSOC'].copy()
+        ASSOCorig.name = 'ASSOC_ORIG'
+        self.update(ASSOCorig, name='ASSOC_ORIG')
+
+        ASSOCnew = self.catalogue['ASSOC'].copy()
+        idxassoc = ~ASSOCnew.isna()
+
+        # Set references first
+        if 'ASSOC_REF' not in self.columns:
+            ASSOCREFnew = Series(np.full(self.catalogue_len, '',
+                                         dtype='U64'),
+                                 name='ASSOC_REF')
+            self.update(ASSOCREFnew, name='ASSOC_REF')
+        else:
+            ASSOCREFnew = self.catalogue['ASSOC_REF'].copy()
+
+        ASSOCREFnew[idxassoc] = ASSOCnew[idxassoc].apply(lambda x:
+                                                         re.split(r'\]', re.split(r'\[', x)[1])[0]
+                                                         if len(re.split(r'\[', x)) > 1 else np.nan)
+
+        # Set values
+        ASSOCnew[idxassoc] = ASSOCnew[idxassoc].apply(lambda x: re.split(r'\[|,|\(|:', x)[0])
+
+        self.update(ASSOCnew)
+        self.update(ASSOCREFnew)
+
+    def parse_type(self):
+        """
+        Parse default string representing source type, extracting (first) value
+        and reference. Multiple values and references currently not supported.
+        """
+
+        if 'TYPE' not in self.columns:
+            return
+
+        # Save original parameter in new column
+        TYPEorig = self.catalogue['TYPE'].copy()
+        TYPEorig.name = 'TYPE_ORIG'
+        self.update(TYPEorig, 'TYPE_ORIG')
+
+        TYPEnew = self.catalogue['TYPE'].copy()
+        idxtype = ~TYPEnew.isna()
+
+        # Set references first
+        if 'TYPE_REF' not in self.columns:
+            TYPEREFnew = Series(np.full(self.catalogue_len, '',
+                                        dtype='U64'),
+                                name='TYPE_REF')
+            self.update(TYPEREFnew, 'TYPE_REF')
+        else:
+            TYPEREFnew = self.catalogue['TYPE_REF'].copy()
+
+        TYPEREFnew[idxtype] = TYPEnew[idxtype].apply(lambda x:
+                                                     re.split(r'\]', re.split(r'\[', x)[1])[0]
+                                                     if len(re.split(r'\[', x)) > 1 else np.nan)
+
+        # Set values
+        TYPEnew[idxtype] = TYPEnew[idxtype].apply(lambda x: re.split(r'\[|,|\(|:', x)[0])
+
+        self.update(TYPEnew)
+        self.update(TYPEREFnew)
```

Comment:
Oh, I think I figured it out:
```suggestion
        self.update(TYPEREFnew, overwrite=True)
```
should do the trick. Actually, I'm not sure why this is required for the 'TYPE_REF' column, which should only have NaNs before this line, but it definitely is required for the `self.update(TYPEnew)` line

Reply:
Thanks for spotting this, I think I've fixed it in fc7efac

Reply:
Hmm, I think that's not quite right. The problem appears to be that `update` ignores all values that are not NaNs (in the original column) when adding new values, so it doesn't change, for example, "HE[aaa+18]" into "HE", so we need to add `overwrite=True` to the `update` lines.

Reply:
Do you think this problem exists in any other functions? Are there any that would be updating an existing value instead of adding a new (non-NaN) value? ... I could probably check myself...

Reply:
Thanks for the fix. I can't think of anywhere else this would be a problem.